onnxruntime/core/codegen/mti/debug/tvm_print.cc

```diff
@@ -0,0 +1,82 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/codegen/mti/debug/tvm_print.h"
+
+#include "core/codegen/common/utils.h"
+#include "core/codegen/mti/common.h"
+#include <topi/detail/extern.h>
+
+namespace onnxruntime {
+namespace tvm_codegen {
+
+TVM_REGISTER_GLOBAL("tvm.contrib.onnxruntime.print")
+    .set_body([](tvm::TVMArgs args, tvm::TVMRetValue* /*ret*/) {
+      DLTensor* X = args[0];
+      DLTensor* Y = args[1];
+
+      DLDataType dtype = X->dtype;
+      std::vector<int64_t> shape;
+      int64_t total_size = 1;
+      for (int i = 0; i < X->ndim; ++i) {
+        shape.push_back(X->shape[i]);
+        total_size *= X->shape[i];
+      }
+
+      // pass X to Y
+      memcpy(static_cast<char*>(Y->data) + Y->byte_offset,
+             static_cast<char*>(X->data) + X->byte_offset,
+             total_size * dtype.bits / 8);
+
+      if (tvm::runtime::TypeMatch(dtype, kDLFloat, 32)) {
+        float* data = reinterpret_cast<float*>(static_cast<char*>(X->data) + X->byte_offset);
+        DumpArray("float tensor:", data, shape);
+      } else if (tvm::runtime::TypeMatch(dtype, kDLInt, 8)) {
```

Comment:
this print function only works on CPU right? I would prefer to put i under nuphar, to avoid in case people thought it is target in-dependent and abuse it.

Reply:
It is for stdio printf support only. FPGA probably won't need it. CUDA, ARM CPU might still have.

Reply:
This is for any CPU devices,  like ARM/X64/PPC. It is target independent, but not for all targets though

---
In reply to: [294438000](https://github.com/microsoft/onnxruntime/pull/888#discussion_r294438000) [](ancestors = 294438000)

Reply:
Will move polynomial implementation outside, and use standard exp/log functions in mti

---
In reply to: [296499584](https://github.com/microsoft/onnxruntime/pull/888#discussion_r296499584) [](ancestors = 296499584,294438000)

Reply:
please document that this debug tool is for cpu only.


---
In reply to: [296499584](https://github.com/microsoft/onnxruntime/pull/888#discussion_r296499584) [](ancestors = 296499584,294438000)